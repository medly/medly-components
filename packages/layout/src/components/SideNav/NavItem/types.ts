import { HTMLProps } from '@medly-components/utils';

export interface NavItemProps extends HTMLProps<HTMLLIElement> {
    /** To be used if rendering like react-router Link */
    to?: string;
    /** Path of the navitem */
    path?: string;
    /** To be used to render Navitem as any component*/
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}

export interface NavItemStyledProps extends HTMLProps<HTMLLIElement> {
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
    isHovered?: boolean;
    isExpanded?: boolean;
    isActive?: boolean;
}
