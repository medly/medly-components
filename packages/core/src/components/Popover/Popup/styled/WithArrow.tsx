import { css } from 'styled-components';
import { PopoverPopupProps } from '../types';

const getPositionWithArrow = ({ placement }: PopoverPopupProps) => {
    switch (placement) {
        case 'top':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateX(-50%);
                    top: 100%;
                    left: 50%;
                }
            `;
        case 'top-start':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    top: 100%;
                    left: 2rem;
                }
            `;
        case 'top-end':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    top: 100%;
                    right: 2rem;
                }
            `;
        case 'right':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateY(-50%) rotate(90deg);
                    top: 50%;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'right-start':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(90deg);
                    top: 2rem;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'right-end':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(90deg);
                    bottom: 2rem;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'bottom':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateX(-50%) rotate(180deg);
                    left: 50%;
                    bottom: 100%;
                }
            `;
        case 'bottom-start':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(180deg);
                    left: 2rem;
                    bottom: 100%;
                }
            `;
        case 'bottom-end':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(180deg);
                    right: 2rem;
                    bottom: 100%;
                }
            `;
        case 'left':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateY(-50%) rotate(270deg);
                    top: 50%;
                    left: calc(100% - 0.3rem);
                }
            `;
        case 'left-start':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(270deg);
                    top: 2rem;
                    left: calc(100% - 0.3rem);
                }
            `;
        case 'left-end':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(270deg);
                    bottom: 2rem;
                    left: calc(100% - 0.3rem);
                }
            `;
    }
};

export const getStyleWithArrow = css<PopoverPopupProps>`
    border-radius: 0.8rem;
    padding: 1.5rem 2rem;

    &::after {
        content: '';
        width: 0;
        height: 0;
        z-index: 4;
        position: absolute;
        border-left: 0.8rem solid transparent;
        border-right: 0.8rem solid transparent;
        border-top: 1rem solid ${({ bgColor }) => bgColor};
    }

    ${getPositionWithArrow}
`;
