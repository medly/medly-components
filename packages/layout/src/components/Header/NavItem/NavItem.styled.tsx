import { breakpoints, media, styled } from '@medly-components/utils';

export type NavItemProps = { isActive?: boolean };
export const NavItem = styled.button<NavItemProps>`
    color: ${({ theme }) => theme.header.navItem.fontColor};
    padding: 0 1.2rem;
    border-radius: 1rem;
    background: ${({ theme, ...props }) => (props.isActive ? theme.header.navItem.activeColor : theme.header.backgroundColor)};
    border: none;
    font-family: inherit;
    text-decoration: none;
    font-size: 1.4rem;
    position: relative;
    height: ${({ theme }) => theme.header.navItem.height.mobile};
    display: grid;
    grid-auto-flow: column;
    grid-column-gap: 1.8rem;
    align-items: center;
    text-align: left;
    grid-auto-columns: max-content;
    cursor: pointer;

    ${({ theme, isActive }) => media(breakpoints(theme.breakpoints).up('L'))`
        background: ${theme.header.backgroundColor};
        grid-column-gap: 0.8rem;
        height: ${theme.header.navItem.height.desktop};
        letter-spacing: -0.04rem;
        font-weight: ${isActive ? theme.font.weights.Medium : theme.font.weights.Regular};
        padding: 0 1.6rem;

        &::before {
            display: ${isActive ? 'block' : 'none'};
            content: '';
            position: absolute;
            width: calc(100% - 1.6rem * 2);
            top: calc((${theme.header.navItem.height.desktop} - ${theme.header.height.desktop}) / 2);
            left: 1.6rem;
            background: ${theme.header.navItem.activeIndicatorColor};
            height: 4px;
            border-radius: 0 0 2px 2px;
        }

        &:hover {
            background-color: ${theme.header.navItem.hoverColor};
        }
    `}

    &:active {
        background-color: ${({ theme }) => theme.header.navItem.activeColor};
    }
`;
