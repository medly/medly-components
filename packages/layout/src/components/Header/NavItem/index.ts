export { NavItem as default } from './NavItem.styled';
