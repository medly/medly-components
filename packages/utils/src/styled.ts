import * as styledComponents from 'styled-components';

const { default: styled, css, createGlobalStyle, keyframes, ThemeProvider, withTheme } = styledComponents;

export { styled, css, createGlobalStyle, keyframes, ThemeProvider, withTheme };
