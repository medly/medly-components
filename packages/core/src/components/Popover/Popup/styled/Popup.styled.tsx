import { fullHeight, fullWidth } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import { Placement } from '../../types';
import { StyledPopoverPopupProps } from '../types';
import { getStyleWithArrow } from './WithArrow';

export const getPosition = ({ placement, distance = '0px' }: { placement: Placement; distance?: string }) => {
    switch (placement) {
        case 'top-start':
            return css`
                bottom: calc(100% + ${distance});
                left: 0%;
            `;
        case 'top':
            return css`
                bottom: calc(100% + ${distance});
                left: 50%;
                transform: translateX(-50%);
            `;
        case 'top-end':
            return css`
                bottom: calc(100% + ${distance});
                right: 0%;
            `;
        case 'right-start':
            return css`
                top: 0%;
                left: calc(100% + ${distance});
            `;
        case 'right':
            return css`
                left: calc(100% + ${distance});
                top: 50%;
                transform: translateY(-50%);
            `;
        case 'right-end':
            return css`
                bottom: 0;
                left: calc(100% + ${distance});
            `;
        case 'bottom-start':
            return css`
                top: calc(100% + ${distance});
                left: 0%;
            `;
        case 'bottom':
            return css`
                top: calc(100% + ${distance});
                left: 50%;
                transform: translateX(-50%);
            `;
        case 'bottom-end':
            return css`
                top: calc(100% + ${distance});
                right: 0%;
            `;
        case 'left-start':
            return css`
                right: calc(100% + ${distance});
                top: 0%;
            `;
        case 'left':
            return css`
                right: calc(100% + ${distance});
                top: 50%;
                transform: translateY(-50%);
            `;
        case 'left-end':
            return css`
                right: calc(100% + ${distance});
                bottom: 0%;
            `;
    }
};

export const PopupStyled = styled('div')<StyledPopoverPopupProps>`
    z-index: 4;
    position: absolute;
    background-color: ${({ bgColor }) => bgColor};

    ${props => props.fullWidth && fullWidth()};
    ${props => props.fullHeight && fullHeight()};

    ${props => getPosition({ placement: props.placement, distance: props.withArrow ? '1.5rem' : props.distance })}
    ${({ withArrow }) => withArrow && getStyleWithArrow}
`;
