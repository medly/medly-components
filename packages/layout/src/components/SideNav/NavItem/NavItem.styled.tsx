import { Text } from '@medly-components/core';
import { SvgIcon } from '@medly-components/icons';
import { defaultTheme } from '@medly-components/theme';
import { getFontStyle } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import { NavItemStyledProps } from './types';

const getStyle = ({ navItem, state }: NavItemStyledProps & { state: 'default' | 'hovered' | 'pressed' | 'active' }) => css`
    background-color: ${navItem.bgColor[state]};
    ${SvgIcon} {
        * {
            fill: ${navItem.icon.color[state]};
        }
    }
    ${Text.Style} {
        color: ${navItem.text.color[state]};
    }
`;
export const NavItemStyled = styled('li').attrs(({ theme: { sideNav } }) => ({ ...sideNav }))<NavItemStyledProps>`
    text-decoration: none;
    box-sizing: border-box;
    display: grid;
    margin: 0 1.2rem;
    border-radius: ${({ navItem }) => navItem.borderRadius};
    box-shadow: ${({ theme, isActive }) => isActive && `0 0 2.4rem ${theme.colors.grey[100]}`};
    grid-template-columns: ${({ openSize, closeSize }) => `calc(${closeSize} - 1.2rem) calc(${openSize} - ${closeSize} - 1.2rem) `};
    justify-items: flex-start;
    align-items: center;
    user-select: none;
    min-height: ${({ navItem }) => navItem.minHeight};
    transition: all 100ms ease-out;
    cursor: pointer;

    ${SvgIcon} {
        overflow: visible;
        border-radius: ${({ navItem }) => navItem.borderRadius};
        padding: ${({ navItem, theme }) => `calc((${navItem.minHeight} - ${theme.icon.sizes[navItem.icon.size].iconSize}) / 2) `};
        font-size: ${({ theme, navItem }) => theme.icon.sizes[navItem.icon.size].iconSize};
    }

    ${Text.Style} {
        text-overflow: ellipsis;
        white-space: nowrap;
        justify-self: left;
        overflow: hidden;
        transition: opacity 200ms ease-out, color 100ms ease-out;
        opacity: ${({ isHovered, isExpanded }) => (isHovered || isExpanded ? 1 : 0)};
        ${({ theme, isActive }) =>
            getFontStyle({ theme, fontVariant: theme.sideNav.navItem.text.textVariant, fontWeight: isActive ? 'Medium' : 'Regular' })}
    }

    ${props => getStyle({ ...props, state: props.isActive ? 'active' : 'default' })};
    &:hover {
        ${props => !props.isActive && getStyle({ ...props, state: 'hovered' })};
    }
    &:active {
        ${props => !props.isActive && getStyle({ ...props, state: 'pressed' })};
    }
`;

NavItemStyled.defaultProps = {
    theme: defaultTheme
};
