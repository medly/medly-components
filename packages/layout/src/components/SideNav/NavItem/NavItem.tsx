import { WithStyle } from '@medly-components/utils';
import { FC, MouseEvent, useCallback, useContext, useMemo } from 'react';
import SideNavContext from '../SideNav.context';
import { NavItemStyled } from './NavItem.styled';
import { NavItemProps } from './types';

export const NavItem: FC<NavItemProps> & WithStyle = props => {
    const { activeItem, activeItemChangeHandler, isHovered, isExpanded } = useContext(SideNavContext),
        { path, onClick, to, ...restProps } = props;

    const isActive = useMemo(() => path === activeItem || to === activeItem, [to, path, activeItem]),
        onClickHandler = useCallback(
            (event: MouseEvent<HTMLLIElement>) => {
                const newItem = path || to;
                event.stopPropagation();
                onClick && onClick(event);
                newItem && activeItemChangeHandler(newItem);
            },
            [onClick, path]
        );

    return <NavItemStyled {...{ ...restProps, isHovered, isExpanded, isActive }} onClick={onClickHandler} />;
};

NavItem.displayName = 'NavItem';
NavItem.Style = NavItemStyled;
