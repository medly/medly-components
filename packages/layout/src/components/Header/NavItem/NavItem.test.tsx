import { render } from '@test-utils';
import { NavItem } from './NavItem.styled';

describe('Header NavItem', () => {
    test.each([true, false])(`should render properly when isActive is set to %p`, isActive => {
        const { container } = render(<NavItem isActive={isActive} />);
        expect(container).toMatchSnapshot();
    });
});
