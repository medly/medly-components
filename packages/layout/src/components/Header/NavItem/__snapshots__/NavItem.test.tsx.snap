// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header NavItem should render properly when isActive is set to false 1`] = `
.c0 {
  padding: 0 1.2rem;
  border-radius: 0.4rem;
  border: none;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.4rem;
  position: relative;
  height: 48px;
  display: grid;
  grid-auto-flow: column;
  grid-column-gap: 1.8rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: left;
  grid-auto-columns: max-content;
  cursor: pointer;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  color: #ffffff;
  background: transparent;
}

.c0 svg * {
  fill: #ffffff;
}

.c0:hover {
  color: #ffffff;
  background: #1F6465;
}

.c0:hover svg * {
  fill: #ffffff;
}

.c0:active {
  color: #ffffff;
  background: #1A5556;
}

.c0:active svg * {
  fill: #ffffff;
}

.c0::after {
  content: '';
  position: absolute;
  display: none;
  background: #3CDBC0;
  width: 0.4rem;
  left: -1rem;
  height: 100%;
}

@media (min-width:1024px) {
  .c0 {
    background: #012040;
    grid-column-gap: 0.8rem;
    height: 56px;
    -webkit-letter-spacing: -0.04rem;
    -moz-letter-spacing: -0.04rem;
    -ms-letter-spacing: -0.04rem;
    letter-spacing: -0.04rem;
    font-weight: 400;
    padding: 0 1.6rem;
  }

  .c0::after {
    bottom: calc((56px - 72px) / 2);
    width: calc(100% - 1.6rem * 2);
    left: 1.6rem;
    height: 0.4rem;
  }
}

<div>
  <button
    class="c0"
  />
</div>
`;

exports[`Header NavItem should render properly when isActive is set to true 1`] = `
.c0 {
  padding: 0 1.2rem;
  border-radius: 0.4rem;
  border: none;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.4rem;
  position: relative;
  height: 48px;
  display: grid;
  grid-auto-flow: column;
  grid-column-gap: 1.8rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: left;
  grid-auto-columns: max-content;
  cursor: pointer;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  color: #3CDBC0;
  background: transparent;
}

.c0 svg * {
  fill: #3CDBC0;
}

.c0:hover {
  color: #ffffff;
  background: #1F6465;
}

.c0:hover svg * {
  fill: #ffffff;
}

.c0:active {
  color: #ffffff;
  background: #1A5556;
}

.c0:active svg * {
  fill: #ffffff;
}

.c0::after {
  content: '';
  position: absolute;
  display: block;
  background: #3CDBC0;
  width: 0.4rem;
  left: -1rem;
  height: 100%;
}

@media (min-width:1024px) {
  .c0 {
    background: #012040;
    grid-column-gap: 0.8rem;
    height: 56px;
    -webkit-letter-spacing: -0.04rem;
    -moz-letter-spacing: -0.04rem;
    -ms-letter-spacing: -0.04rem;
    letter-spacing: -0.04rem;
    font-weight: 600;
    padding: 0 1.6rem;
  }

  .c0::after {
    bottom: calc((56px - 72px) / 2);
    width: calc(100% - 1.6rem * 2);
    left: 1.6rem;
    height: 0.4rem;
  }
}

<div>
  <button
    class="c0"
  />
</div>
`;
